book/src/java_objects_in_rust.md

```diff
@@ -0,0 +1,75 @@
+# Java Objects in Rust
+
+At this point, suppose we’ve concluded that we need to run `counter_increment`
+on a large number of objects at once. We might add an API to the Java side like
+so:
+
+
+```java
+class NativeAPI {
+    ...
+    static native void increment_all(NativeCounter[] counters)
+}
+```
+
+On the Rust side, our corresponding signature will be:
+
+```rust
+#[no_mangle]
+pub extern "system" fn Java_com_github_jni_1rs_jnibook_NativeAPI_increment_all(
+    env: JNIEnv,
+    _class: JClass,
+    counters: jobjectArray,
+) {
+    let length = env.get_array_length(pointers)
+        .expect("TODO: proper error handling");
+    for index in 0..length {
+        let item = env.get_object_array_element(counters)
+            .expect("TODO: proper error handling");
+        let pointer = env.get_field(item, "ptr", "Z")
+            .expect("TODO: proper error handling");
+        // TODO: Obtain the Rust object and invoke it.
+    }
+}
+```
+
+## Exercises
+
+1. Fill in the code to invoke `increment()` on the underlying Rust object.
+2. If this code is run with a large enough object array, it will crash. That’s
+   because the call to `get_object_array_element` creates a local reference, and
+   only a small number of local references are permitted at a time (the exact
+   number may vary based on your OS). Fix the code using one of the following
+   approaches:
+    1. Releasing the local reference when you’re done with it, using
+       [`delete_local_ref`](https://docs.rs/jni/0.15.0/jni/struct.JNIEnv.html#method.delete_local_ref),
+       as you might in C++.
+    2. Give each iteration its own local reference frame using
+       [`with_local_frame`](https://docs.rs/jni/0.15.0/jni/struct.JNIEnv.html#method.with_local_frame),
+       so that all references within that frame are released by the next
+       iteration.
+    3. Use the
+       [`AutoLocal`](https://docs.rs/jni/0.15.0/jni/struct.JNIEnv.html#method.auto_local)
+       helper implemented within the JNI crate, so you don’t have to manually
+       track the usage of that reference at all.
```

Comment:
I think a rule of thumb is that if you create more than a couple of AutoLocals,
       with_local_frame will be more efficient (I think we have benchmarks for that in `//benches/`),
       therefore it's not like AutoLocals are always better. If docs are unclear on that, it might be
       better to update them instead of repeating them here.

Reply:
The `AutoLocal` docs (https://docs.rs/jni/0.18.0/jni/objects/struct.AutoLocal.html) don't currently mention `with_local_frame`. I'd also expect `with_local_frame` to be more efficient, and think that's worth mentioning in the rustdocs.

Is there any way I can add this as a task to the project board? I don't think I have permissions to manipulate it.

Reply:
> Is there any way I can add this as a task to the project board? I don't think I have permissions to manipulate it.

I've sent a collaborator invitation earlier, but see it is still pending. Likely, it must be somewhere in the mail. I think when you accept you must be able to modify the board.

Reply:
Aha, found it. Thank you :)